### ISSUE TITLE
cmd/compile: optimize struct copy

### ISSUE THREAD

**Issue by tdewolff** [golang/go#12841]
Go1.5.1 windows/amd64

It seems that manually copying all the fields of a struct to a new location is faster than copying through assignment (which invokes `memmove`).

```
BenchmarkMoveSmallStruct-8      2000000000               0.62 ns/op
BenchmarkSetSmallStruct-8       2000000000               0.50 ns/op
BenchmarkMoveBigStruct-8        1000000000               2.68 ns/op
BenchmarkSetBigStruct-8         1000000000               2.20 ns/op
```

Using
``` go
package main

import "testing"

type Small struct {
	a int
}

type Big struct {
	a, b, c, d, e, f, g, h int
}

////////////////

func moveSmall(item *Small) {
	*item = Small{0}
}

func setSmall(item *Small) {
	item.a = 0
}

func moveBig(item *Big) {
	*item = Big{0, 1, 2, 3, 4, 5, 6, 7}
}

func setBig(item *Big) {
	item.a, item.b, item.c, item.d, item.e, item.f, item.g, item.h = 0, 1, 2, 3, 4, 5, 6, 7
}

////////////////

func BenchmarkMoveSmallStruct(b *testing.B) {
	item := Small{0}
	for i := 0; i < b.N; i++ {
		moveSmall(&item)
	}
}

func BenchmarkSetSmallStruct(b *testing.B) {
	item := Small{0}
	for i := 0; i < b.N; i++ {
		setSmall(&item)
	}
}

func BenchmarkMoveBigStruct(b *testing.B) {
	item := Big{0, 1, 2, 3, 4, 5, 6, 7}
	for i := 0; i < b.N; i++ {
		moveBig(&item)
	}
}

func BenchmarkSetBigStruct(b *testing.B) {
	item := Big{0, 1, 2, 3, 4, 5, 6, 7}
	for i := 0; i < b.N; i++ {
		setBig(&item)
	}
}
```

I think the assignment of the struct first allocates the struct on the stack and then `memmoves` it to the destination. The struct can be set on the destination immediately.

**Comment by gyf19** [comment]
+1

**Comment by randall77** [comment]
Just took a look at this on the SSA branch.
You're right, there are two copies in the bad case, global data -> stack and stack -> destination.

The copies themselves are ok, if you do *dst = *src vs.  dst.a, ... = src.a, ... the copy (using duffcopy) is slightly faster.  It's the x2 that is the problem.

There doesn't seem to be an easy fix for this.  For instance, if you do:
```
func moveBigX(item *Big, x int) {
	*item = Big{0, 1, 2, x, 4, 5, 6, 7}
}
```
then the intermediate stack copy is necessary so you can modify the stack data before writing it.  Direct assignments are still better in this case, but if the struct gets large enough the code size becomes prohibitive.

A rule like:
(Move dst inter (Move inter src mem)) -> (Move dst src mem)
might work.  It's not quite that simple, as we have to make sure the intermediate memory isn't used by anyone else.  And drop the inter variable if it is no longer used.  And probably lots of other conditions I'm missing.

Long story short, thinking about it.  Should be doable for the completely constant case.
